x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/persistence/JobConfigProvider.java

```diff
@@ -0,0 +1,396 @@
+import org.elasticsearch.common.xcontent.XContentType;
+import org.elasticsearch.index.query.BoolQueryBuilder;
+import org.elasticsearch.index.query.QueryBuilder;
+import org.elasticsearch.index.query.TermQueryBuilder;
+import org.elasticsearch.index.query.TermsQueryBuilder;
+import org.elasticsearch.index.query.WildcardQueryBuilder;
+import org.elasticsearch.search.SearchHit;
+import org.elasticsearch.search.builder.SearchSourceBuilder;
+import org.elasticsearch.xpack.core.ml.job.config.Job;
+import org.elasticsearch.xpack.core.ml.job.config.JobUpdate;
+import org.elasticsearch.xpack.core.ml.job.persistence.AnomalyDetectorsIndex;
+import org.elasticsearch.xpack.core.ml.job.persistence.ElasticsearchMappings;
+import org.elasticsearch.xpack.core.ml.utils.ExceptionsHelper;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import static org.elasticsearch.xpack.core.ClientHelper.ML_ORIGIN;
+import static org.elasticsearch.xpack.core.ClientHelper.executeAsyncWithOrigin;
+
+/**
+ * This class implements CRUD operation for the
+ * anomaly detector job configuration document
+ */
+public class JobConfigProvider extends AbstractComponent {
+
+    public static String ALL = "_all";
+
+    private final Client client;
+
+    public JobConfigProvider(Client client, Settings settings) {
+        super(settings);
+        this.client = client;
+    }
+
+    /**
+     * Persist the anomaly detector job configuration to the configuration index.
+     * It is an error if an job with the same Id already exists - the config will
+     * not be overwritten.
+     *
+     * @param job The anomaly detector job configuration
+     * @param listener Index response listener
+     */
+    public void putJob(Job job, ActionListener<IndexResponse> listener) {
+        try (XContentBuilder builder = XContentFactory.jsonBuilder()) {
+            XContentBuilder source = job.toXContent(builder, ToXContent.EMPTY_PARAMS);
+            IndexRequest indexRequest =  client.prepareIndex(AnomalyDetectorsIndex.configIndexName(),
+                    ElasticsearchMappings.DOC_TYPE, Job.documentId(job.getId()))
+                    .setSource(source)
+                    .setOpType(DocWriteRequest.OpType.CREATE)
+                    .request();
+
+            executeAsyncWithOrigin(client, ML_ORIGIN, IndexAction.INSTANCE, indexRequest, listener);
+
+        } catch (IOException e) {
+            listener.onFailure(new ElasticsearchParseException("Failed to serialise job with id [" + job.getId() + "]", e));
+        }
+    }
+
+    /**
+     * Get the anomaly detector job specified by {@code jobId}.
+     * If the job is missing a {@code ResourceNotFoundException} is returned
+     * via the listener.
+     *
+     * @param jobId The job ID
+     * @param jobListener Job listener
+     */
+    public void getJob(String jobId, ActionListener<Job.Builder> jobListener) {
```

Comment:
I think this should be returning the built job.

Reply:
I wanted to avoid the current situation we have where we cannot change the validation as building the job will throw. Getting an old job on an upgraded node where the validation has changed (due to some bug say) would be impossible if we validate here so you couldn't even view the job. This class is just for config index operations the validation can be done elsewhere at the appropriate place.